### PR TITLE
Final project

### PR DESCRIPTION
**Pull request by NapoIeonchik**
**Vodafone Music (визначення, чи підпишеться абонент на послугу Music), опис колонок у файлі Feature_dictionary.xlsx**
Обрав задачу із визначенням чи підпишеться абонент на послугу Music

image
Бачимо що у датасеті присутні 460 ознак.
Подивившись у пояснюючий файл бачимо що багато ознак не мають жодного зв'язку , такі як дзвінки, смс, деякі грошові витрати
Також приберемо ці ознак в яких не має 50 відсотків значень(вони майже пусті), інші пусті значення заповнему медіаною.

```
 count_url_category_13    99.308571
 count_url_category_15    98.840000
 vol_app_3                98.805714
 count_app_3              98.805714
 count_url_category_12    98.194286
                            ...    
 block_flag                0.000000
 lt                        0.000000
 tp_flag                   0.000000
 os_category               0.000000
 target                    0.000000
 Length: 460, dtype: float64
 ######## Кількість ознак що мають малу заповненість
 62
```
Таким чином приходимо до 104 ознак.
```
Columns: 104 entries, device_type to target
dtypes: float64(96), int64(8)
```
image
image
image

Подивимось на кореляції із таргет змінною
```
target                1.000000
content_count_m1      0.139147
content_count_m3      0.124602
content_count_m2      0.115864
count_gift_type_3     0.066235
                        ...   
inact_days_count           NaN
rr_act_type_7              NaN
service_8_count_m1         NaN
service_8_count_m2         NaN
service_8_count_m3         NaN
Name: target, Length: 104, dtype: float64
```
Бачимо що кореляційний зв'язок є досить невеликим.
Вибрані ознаки для моделі: ['content_count_m1', 'content_count_m3', 'content_count_m2', 'lt']

Щодо моделей, почнемо із DecisionTreeClassifier

Precision: 0.15897435897435896
Recall: 0.18787878787878787
F1 score: 0.17222222222222222
ROC AUC score: 0.5533955126044126
Accuracy score: 0.8637714285714285

Через те, що таргет змінна є незбалансованою будемо дивитись на результат метрики F1

З налаштуванням через GridSearchCV
Precision: 0.2151394422310757
Recall: 0.16363636363636364
F1 score: 0.18588640275387264
ROC AUC score: 0.5574671311383301
Accuracy score: 0.8918857142857143
Best parameters found:  {'criterion': 'entropy', 'max_depth': None, 'max_features': None, 'min_samples_leaf': 10, 'min_samples_split': 2}
Precision: Значення precision (від 0.16 до 0.22) вказує на те, що модель не дуже ефективно передбачає позитивний клас, маючи багато помилкових спрацьовувань.
Recall: Значення recall (від 0.16 до 0.19) показує, що модель пропускає велику кількість справжніх позитивних прикладів.
F1 Score: F1-міра (0.17–0.18) демонструє збалансованість між precision і recall, але показник залишається низьким, що вказує на слабку продуктивність для позитивного класу.
ROC AUC Score: Значення (0.55) трохи вище випадкового вгадування (0.5), що свідчить про слабку роздільну здатність моделі.
Accuracy: Високе значення (86–89%) говорить про те, що модель добре справляється із негативним класом, але через дисбаланс даних це не є показником загальної якості.
image
image

Далі перейдемо до kNN:
Precision: 0.08571428571428572
Recall: 0.00909090909090909
F1 score: 0.01643835616438356
ROC AUC score: 0.5005899539274076
Accuracy score: 0.9179428571428572

Оптимізована модель:
Precision: 0.0743801652892562
Recall: 0.02727272727272727
F1 score: 0.039911308203991136
ROC AUC score: 0.49979211147319924
Accuracy score: 0.9010285714285714
[[3933  112]
 [ 321    9]]
Best parameters found:  {'algorithm': 'auto', 'metric': 'manhattan', 'n_neighbors': 3, 'weights': 'distance'}
image

Попри оптимізацію гіперпараметрів, метрики Precision, Recall і F1 Score залишилися дуже низькими.
Accuracy трохи знизилась, що може вказувати на спробу моделі краще враховувати позитивний клас, але цього недостатньо.
ROC AUC Score залишився близьким до 0.5, що вказує на нездатність моделі ефективно розрізняти класи.

KNN не підходить для задач, де є сильний дисбаланс класів, оскільки він враховує кількість сусідів незалежно від частоти класів.

Модель AdaBoost:
Precision: 0.37142857142857144
Recall: 0.11818181818181818
F1 score: 0.1793103448275862
ROC AUC score: 0.5509326890661872
Accuracy score: 0.9184

Precision значно вищий, ніж у KNN чи Decision Tree, що свідчить про кращу здатність правильно ідентифікувати позитивний клас.
Однак Recall залишається низьким, що вказує на пропуск багатьох позитивних прикладів.
F1 Score все ще на низькому рівні, але показує тенденцію до покращення порівняно з попередніми моделями.
ROC AUC Score трохи вищий за випадковий розподіл (0.5), але різниця незначна.

Оптимізація n_estimators:
Precision: 0.3711340206185567
Recall: 0.10909090909090909
F1 score: 0.1686182669789227
ROC AUC score: 0.5470052814923025
Accuracy score: 0.9188571428571428
Кращі гіперпараметри:n_estimators: 250
Незначне зниження Recall і F1 Score після оптимізації може бути пов’язане з переобчисленням ваг та покращенням стабільності моделі.
Precision залишився стабільним, що свідчить про збереження здатності коректно класифікувати позитивний клас.
image

Оптимізація learning_rate:
Precision: 0.32413793103448274
Recall: 0.14242424242424243
F1 score: 0.19789473684210523
ROC AUC score: 0.5590984005693523
Accuracy score: 0.9129142857142857
image
Кращі гіперпараметри: learning_rate: ≈ 0.01
Recall значно покращився, хоча Precision знизився. Це призвело до невеликого зростання F1 Score.
Показник ROC AUC також трохи зріс, що свідчить про покращення здатності моделі розрізняти класи.
Accuracy не зазнала суттєвих змін, оскільки домінуючий клас "0" як і раніше впливає на метрику.
Загальні висновки для AdaBoost:
Переваги:
AdaBoost демонструє значно кращу продуктивність у порівнянні з KNN та Decision Tree.
Модель краще працює з дисбалансом класів завдяки динамічному коригуванню ваг на кожній ітерації.
Недоліки:
Recall все ще залишається низьким, що означає недостатню ідентифікацію позитивного класу.
F1 Score хоча й найвищий серед моделей, все ще знаходиться на невисокому рівні.

Модель XGBoost:
Precision: 0.39473684210526316
Recall: 0.09090909090909091
F1 score: 0.14778325123152708
ROC AUC score: 0.5397685133161029
Accuracy score: 0.9209142857142857
XGBoost у своїй базовій формі демонструє непоганий Precision, однак Recall залишається низьким.
F1 Score та ROC AUC Score вказують на лише невелике покращення порівняно з AdaBoost, але точність класифікації домінуючого класу "0" ще сильно впливає на Accuracy.

Оптимізація n_estimators:
Precision: 0.4305555555555556
Recall: 0.09393939393939393
F1 score: 0.154228855721393
ROC AUC score: 0.5419017118028243
Accuracy score: 0.9222857142857143
Кращі гіперпараметри: n_estimators: 250
Легка оптимізація кількості дерев призвела до збільшення Precision, але Recall залишається незмінно низьким.
Незначне підвищення F1 Score та ROC AUC свідчить, що модель краще обробляє тренувальні дані.
image

Оптимізація max_depth:
Precision: 0.336734693877551
Recall: 0.1
F1 score: 0.1542056074766355
ROC AUC score: 0.5419653893695922
Accuracy score: 0.9172571428571429
Кращі гіперпараметри: max_depth: 4
Precision знижується, але Recall трохи покращується, що свідчить про зміщення фокусу моделі на більш збалансоване представлення обох класів.
Загальний F1 Score залишився незмінним, а Accuracy трохи знизився через більший акцент на менш частий клас "1".
image
Загальні висновки для XGBoost:
Переваги:
XGBoost забезпечує трохи кращі результати, ніж AdaBoost, зокрема Precision.
Гнучкість в налаштуванні гіперпараметрів дозволяє покращувати продуктивність.
Недоліки:
Recall залишається низьким навіть після оптимізації, що є критичним для виявлення позитивного класу.
Значний дисбаланс класів впливає на метрики, особливо на ROC AUC.

### CODE REVIEW

**Comment by NapoIeonchik** [comment]
Target distribution: 
0 - 92.342857
1 - 7.657143
